### PR TITLE
Issue #2693 Implement PtNDArrayEx.multiBoxPrior with validation

### PR DESCRIPTION
**Pull request by juliangamble**
## Description ##

The big goal here is to implement `PtNDArrayEx.multiBoxPrior()` (currently throws `NotImplementException`) so that you can run `PikachuTraining` with Pytorch and so you can run SSD Training on M1 Macs. 

This is discussed in: https://github.com/deepjavalibrary/djl/issues/2693

The big idea is to prove this works by writing matching unit tests for `MxNDAarrayEx.multiBoxPrior()` and comparing the results. 

The implementation for this is derived from:
https://github.com/apache/mxnet/blob/master/src/operator/contrib/multibox_prior.cc
and
https://github.com/apache/mxnet/blob/a720b15b5fa011a9610dfaeabf0792443c6abec5/src/operator/contrib/multibox_prior-inl.h#L116

This implementation is based on implementing the unit test on the C++ side (see unit test in pull request there) and comparing the steps internally.

### CODE REVIEW

File: engines/mxnet/mxnet-engine/src/test/java/ai/djl/mxnet/engine/MxNDArrayExTest.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+import ai.djl.ndarray.internal.NDArrayEx;
+import ai.djl.ndarray.types.DataType;
+import ai.djl.ndarray.types.Shape;
+import org.testng.annotations.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+public class MxNDArrayExTest {
```
Comment (frankfliu) [inline engines/mxnet/mxnet-engine/src/test/java/ai/djl/mxnet/engine/MxNDArrayExTest.java:16]: We don't need this test, we have integration tests for multiBoxPrior: https://github.com/deepjavalibrary/djl/blob/master/integration/src/main/java/ai/djl/integration/tests/modality/cv/MultiBoxPriorTest.java#L30

The integration test covers both MXNet and PyTorch

File: engines/pytorch/pytorch-engine/src/test/java/ai/djl/pytorch/engine/PtNDArrayExTest.java [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+
+
+public class PtNDArrayExTest {
+    public static void main(String[] args) {
+        PtNDArrayExTest ptNDArrayExTest = new PtNDArrayExTest();
+        ptNDArrayExTest.testMultiBoxPrior();
+    }
+
+    @Test
+    public void testMultiBoxPrior() {
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/test/java/ai/djl/pytorch/engine/PtNDArrayExTest.java:22]: You can use MultiBoxPriorTest to test for PyTorch as well.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 3]
```diff
@@ -26,6 +26,8 @@
 
 import java.util.List;
 
+import static ai.djl.pytorch.jni.JniUtils.sqrt;
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:29]: Remove unused import

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 4]
```diff
@@ -694,7 +696,59 @@ public NDList multiBoxPrior(
             List<Float> steps,
             List<Float> offsets,
             boolean clip) {
-        throw new UnsupportedOperationException("Not implemented");
+        //Based on https://github.com/apache/mxnet/blob/a720b15b5fa011a9610dfaeabf0792443c6abec5/src/operator/contrib/multibox_prior.cc
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:699]: Please format java code with:

```
./gradlew fJ
```

See: https://docs.djl.ai/docs/development/development_guideline.html#coding-conventions

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 5]
```diff
@@ -694,7 +696,59 @@ public NDList multiBoxPrior(
...
+
+        NDManager ndManager = array.getManager().getParentManager();
+
+        Float step_x   = steps.get(1);
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:703]: Please follow our coding convention, we use google AOSP format

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 6]
```diff
@@ -694,7 +696,59 @@ public NDList multiBoxPrior(
...
+        int num_ratios = ratios.size();
+        int count = 0;
+        float[][] out = new float[num_sizes][num_ratios];
+
+        //Based on https://github.com/apache/mxnet/blob/a720b15b5fa011a9610dfaeabf0792443c6abec5/src/operator/contrib/multibox_prior-inl.h#L116
+
+        PtNDArray input = array;
+        //int in_height         = in_data[mboxprior_enum::kData].size(2);
+        //int in_width          = in_data[mboxprior_enum::kData].size(3);
+        int in_height         = input.getInt(2);
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:715]: This should be shape, right?

```
input.getShape().get(2)
```

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 7]
```diff
@@ -694,7 +696,59 @@ public NDList multiBoxPrior(
...
+        //Based on https://github.com/apache/mxnet/blob/a720b15b5fa011a9610dfaeabf0792443c6abec5/src/operator/contrib/multibox_prior.cc
+
+        NDManager ndManager = array.getManager().getParentManager();
+
+        Float step_x   = steps.get(1);
+        Float step_y   = steps.get(0);
+        int num_sizes  = sizes.size();
+        int num_ratios = ratios.size();
+        int count = 0;
+        float[][] out = new float[num_sizes][num_ratios];
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:708]: the out dimension is wrong, please make sure your implementation is correct

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 8]
```diff
@@ -694,7 +696,59 @@ public NDList multiBoxPrior(
             List<Float> steps,
             List<Float> offsets,
             boolean clip) {
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:698]: clip is not been used in your implementaiton

**Review by frankfliu — CHANGES_REQUESTED** [review]

File: engines/mxnet/mxnet-engine/src/test/java/ai/djl/mxnet/engine/MxNDArrayExTest.java (hunk 1 above)
Comment (juliangamble) [reply engines/mxnet/mxnet-engine/src/test/java/ai/djl/mxnet/engine/MxNDArrayExTest.java]: Cool - deleted.

File: engines/pytorch/pytorch-engine/src/test/java/ai/djl/pytorch/engine/PtNDArrayExTest.java (hunk 2 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/test/java/ai/djl/pytorch/engine/PtNDArrayExTest.java]: Cool - deleted.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 3 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Cool - done.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 4 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Cool - done.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 5 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Could you help me out here? It's a big document. Are you referring to a particular rule?

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 6 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Thanks - that's helpful. Fixed.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 7 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Agree this is wrong. Comes out very clearly in the unit/integration test: `MultiBoxPriorTest`. I am asking for help. 
The code here
https://github.com/apache/mxnet/blob/master/src/operator/contrib/multibox_prior.cc#L53
refers to a two dimensional array. 

In addition, `manager.create` takes a two dimensional array. 

Can you help me out with where in the process you add the third dimension?

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 8 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:696]: Thanks - agree it is not used. I'm asking for help here - and reading the baseline implementation in mxnet. 

Looking at the baseline code:
https://github.com/apache/mxnet/blob/master/src/operator/contrib/multibox_prior.cc#L53
We don't see `clip` used there either. (Correct me if I'm wrong)

We do see a reference to `clip` in the interface for this java method so we have to implement it as part of our class. 
https://github.com/deepjavalibrary/djl/blob/master/api/src/main/java/ai/djl/ndarray/internal/NDArrayEx.java#L601
 
There are hints of an implementation using `clip` in old mxnet code:
https://github.com/apache/mxnet/blob/a720b15b5fa011a9610dfaeabf0792443c6abec5/src/operator/contrib/multibox_prior-inl.h#L128
But it appears to have been removed.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 9]
```diff
@@ -694,7 +694,62 @@ public NDList multiBoxPrior(
...
+        // Based on
+        // https://github.com/apache/mxnet/blob/a720b15b5fa011a9610dfaeabf0792443c6abec5/src/operator/contrib/multibox_prior-inl.h#L116
+
+        Shape input = array.getShape();
+        // int in_height         = in_data[mboxprior_enum::kData].size(2);
+        // int in_width          = in_data[mboxprior_enum::kData].size(3);
+        int in_height = (int) input.get(2);
+        int in_width = (int) input.get(3);
+
+        float[][] out = new float[(in_height + 1) * (in_width + 1) * (num_sizes + 1) * 2][4];
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:717]:
```suggestion
        float[][] out = new float[in_height * in_width * num_sizes * 2][4];
```

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 10]
```diff
@@ -694,7 +694,62 @@ public NDList multiBoxPrior(
...
+        int num_ratios = ratios.size();
+        int count = 0;
+
+        // Based on
+        // https://github.com/apache/mxnet/blob/a720b15b5fa011a9610dfaeabf0792443c6abec5/src/operator/contrib/multibox_prior-inl.h#L116
+
+        Shape input = array.getShape();
+        // int in_height         = in_data[mboxprior_enum::kData].size(2);
+        // int in_width          = in_data[mboxprior_enum::kData].size(3);
+        int in_height = (int) input.get(2);
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:714]:
```suggestion
        int in_height = Math.toIntExact(input.get(2));
```

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 11]
```diff
@@ -694,7 +694,62 @@ public NDList multiBoxPrior(
...
+                    float h = size / ratioLocal / 2;
+                    out[count][0] = center_x - w; // xmin
+                    out[count][1] = center_y - h; // ymin
+                    out[count][2] = center_x + w; // xmax
+                    out[count][3] = center_y + h; // ymax
+                    ++count;
+                }
+            }
+        }
+        NDArray ndArray = ndManager.create(out);
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:750]:
```suggestion
        NDArray ndArray = ndManager.create(out).expendDims(0);
```

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 7 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Thanks for the help.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 9 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Thankyou. Done.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 10 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Thankyou. Done

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 11 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: That's what I was looking for!! 
Thankyou. Done.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 12]
```diff
@@ -694,7 +694,62 @@ public NDList multiBoxPrior(
...
+        NDList outResult = new NDList(ndArray);
```
[the remark this answers is not in the document]
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Thanks. Done.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 13]
```diff
@@ -694,7 +694,54 @@ public NDList multiBoxPrior(
             List<Float> steps,
             List<Float> offsets,
             boolean clip) {
-        throw new UnsupportedOperationException("Not implemented");
+
+        NDManager ndManager = array.getManager().getParentManager();
+
+        Float stepX = steps.get(1);
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:700]:
```suggestion
        float stepX = steps.get(1);
```

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 14]
```diff
@@ -694,7 +694,54 @@ public NDList multiBoxPrior(
...
+        Float stepY = steps.get(0);
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:701]:
```suggestion
        float stepY = steps.get(0);
```

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 15]
```diff
@@ -694,7 +694,54 @@ public NDList multiBoxPrior(
             List<Float> steps,
             List<Float> offsets,
             boolean clip) {
-        throw new UnsupportedOperationException("Not implemented");
+
+        NDManager ndManager = array.getManager().getParentManager();
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:698]:
```suggestion
        NDManager ndManager = array.getManager();
```

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java [hunk 16]
```diff
@@ -694,7 +694,54 @@ public NDList multiBoxPrior(
...
+                    float h = size / ratioLocal / 2;
+                    out[count][0] = centerX - w; // xmin
+                    out[count][1] = centerY - h; // ymin
+                    out[count][2] = centerX + w; // xmax
+                    out[count][3] = centerY + h; // ymax
+                    ++count;
+                }
+            }
+        }
+        NDArray ndArray = ndManager.create(out).expandDims(0);
```
Comment (frankfliu) [inline engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:743]: I didn't look into your algorithm, but seems the final output is different from MXNet. Can you debug why it's different?

Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:743]: There is some discussion of this here:
https://github.com/deepjavalibrary/djl/issues/2693

I'd love to debug it on the mxnet side, but this is non-trivial.  I'm working on a method to do so, but would appreciate some pointers.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 15 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Thankyou. This has been updated.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 14 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Thankyou. This has been updated.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 13 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java]: Thankyou. This has been updated.

File: engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java (hunk 16 above)
Comment (juliangamble) [reply engines/pytorch/pytorch-engine/src/main/java/ai/djl/pytorch/engine/PtNDArrayEx.java:749]: Thanks @frankfliu I found the bug and put in a fix. 

For me the first element in the result is `-0.09902344` in my implementation and mxnet. 
From your perspective is this now returning the correct values?

**Review by frankfliu — APPROVED** [review]
Thanks so much for your contribution.

**Pull request merged**